vulcan/lib/dependency_manager.rb

```diff
@@ -101,9 +105,13 @@ def latest_image(image_name)
     end
 
     def image_exists?(image_name)
+      key_check_result = @existing_images_cache[image_name]
+      return key_check_result unless key_check_result.nil?
+
       # Run a docker command to check that the sha or image exists.
       exists = `docker manifest inspect #{image_name} > /dev/null`
```

Comment:
Part of the performance likely also is the required spawning and process management overhead for the operation.  I believe docker hub's api is a simple HTTP interface; can we figure out what URL is required to ping the existance of an image?

Reply:
Maybe just followed up PR work

Reply:
Oh, I see, that makes sense. Yeah, I can add that as a follow on task to change from using the docker CLI to pinging the API directly. The nice thing about the CLI would be that it would stay updated (hopefully) regarding any underlying API changes...

Reply:
That's true, although it's also _sensitive_ to changes in the CLI, including flags and how authentication works, too.  It's also more tricky to consistently test as a result, compared to an HTTP mock.

Reply:
:+1:

Reply:
That's probably the right call, and it might be contributing to why production is slow.

__not all nodes in production are authenticated__ with dockerhub.  leader nodes do the image pulling and propagation in many cases.  I believe in fact that unauthenticated access IS available but behind rate limiting.

Reply:
Oooh, interesting. I'll have to try some additional endpoints, because so far everything I do manually returns 401 asking me to authenticate...

Reply:
that might have actually changed recently now that I recall so I may be wrong lol

Reply:
The v1 API seems open to unauthenticated access, but I can't quite tell if the manifest information is available through it (image manifest seems like a v2 thing?)...

Reply:
Might be some way to try pulling a given SHA directly, and seeing what happens... :shrug: